renderer/GLSLProgram.cpp

```diff
@@ -0,0 +1,334 @@
+	return glslProgram;
+}
+
+void GLSLProgramLoader::LoadAndAttachShader( GLint shaderType, const char *sourceFile, const idDict &defines ) {
+	if( program == 0 ) {
+		common->Warning( "Tried to attach shader to an already linked program\n" );
+		return;
+	}
+
+	GLuint shader = CompileShader( shaderType, sourceFile, defines );
+	if( shader != 0) {
+		qglAttachShader( program, shader );
+		// won't actually be deleted until the program it's attached to is deleted
+		qglDeleteShader( shader );
+	}
+}
+
+std::string ReadFile( const char *sourceFile ) {
+	void *buf = nullptr;
+	int len = fileSystem->ReadFile( sourceFile, &buf );
+	if( buf == nullptr ) {
+		common->Warning( "Could not open shader file %s", sourceFile );
+		return "";
+	}
+	std::string contents( static_cast< char* >( buf ), len );
+	fileSystem->FreeFile( buf );
+
+	return contents;
+}
+
+/**
+ * Resolves include statements in GLSL source files.
+ * Note that the parsing is primitive and not context-sensitive. It will not respect multi-line comments
+ * or conditional preprocessor blocks, so keep includes simple in the source files!
+ * 
+ * Include directives should look like this:
+ * 
+ * #pragma tdm_include "somefile.glsl" // optional comment
+ */
+void ResolveIncludes( std::string &source, std::vector<std::string> &includedFiles ) {
+	static const std::regex includeRegex( R"regex(^[ \t]*#[ \t]*pragma[ \t]+tdm_include[ \t]+"(.*)"[ \t]*(?:\/\/.*)?\r?$)regex" );
+
+	unsigned int currentFileNo = includedFiles.size() - 1;
+	unsigned int totalIncludedLines = 0;
+
+	std::smatch match;
+	while( std::regex_search( source, match, includeRegex ) ) {
+		std::string fileToInclude( match[ 1 ].first, match[ 1 ].second );
+		if( std::find( includedFiles.begin(), includedFiles.end(), fileToInclude ) == includedFiles.end() ) {
+			int nextFileNo = includedFiles.size();
+			std::string includeContents = ReadFile( fileToInclude.c_str() );
+			includedFiles.push_back( fileToInclude );
+			ResolveIncludes( includeContents, includedFiles );
+
+			// also add a #line instruction at beginning and end of include so that
+			// compile errors are mapped to the correct file and line
+			// unfortunately, #line does not take an actual filename, but only an integral reference to a file :(
+			unsigned int currentLine = std::count( source.cbegin(), match[ 0 ].first, '\n' ) + 1 - totalIncludedLines;
+			std::string includeBeginMarker = "#line 0 " + std::to_string( nextFileNo ) + '\n';
+			std::string includeEndMarker = "\n#line " + std::to_string( currentLine ) + ' ' + std::to_string( currentFileNo );
+			totalIncludedLines += std::count( includeContents.begin(), includeContents.end(), '\n' ) + 2;
+
+			// replace include statement with content of included file
+			source.replace( match[ 0 ].first, match[ 0 ].second, includeBeginMarker + includeContents + includeEndMarker );
+		} else {
+			source.replace( match[ 0 ].first, match[ 0 ].second, "// already included " + fileToInclude );
```

Comment:
BTW, non-guarded includes allow [a useful trick](https://stackoverflow.com/a/17670232/556899) in C-like language. Let's hope no one will ever want this in GLSL shaders =)))

Reply:
Well, that's the point where we'd have to actually look at a full preprocessor solution ;)